mtec/mtec.cpp

```diff
@@ -0,0 +1,155 @@
+* warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+* Lesser General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public License
+* along with this project. If not, see <https://www.gnu.org/licenses/>.
+*
+* For any further details and any questions please contact us under
+* contact@nymea.io or see our FAQ/Licensing Information on
+* https://nymea.io/license/faq
+*
+* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+#include "mtec.h"
+#include "extern-plugininfo.h"
+#include "mtechelpers.h"
+
+#include <cstring>
+
+MTec::MTec(const QHostAddress &address, QObject *parent) :
+    QObject(parent),
+    m_hostAddress(address)
+{
+    m_modbusMaster = new ModbusTCPMaster(address, 502, this);
+
+    if (m_modbusMaster) {
+        qCDebug(dcMTec()) << "created ModbusTCPMaster";
+        m_modbusMaster->connectDevice();
+
+        connect(m_modbusMaster, &ModbusTCPMaster::receivedHoldingRegister, this, &MTec::onReceivedHoldingRegister);
+        connect(m_modbusMaster, &ModbusTCPMaster::readRequestError, this, &MTec::onModbusError);
+        connect(m_modbusMaster, &ModbusTCPMaster::writeRequestError, this, &MTec::onModbusError);
+    }
+}
+
+MTec::~MTec()
+{
+    if (m_modbusMaster) {
+        delete m_modbusMaster;
+    }
+}
+
+void MTec::onModbusError()
+{
+    qCDebug(dcMTec()) << "MTec: Received modbus error";
+
+    if (m_info != nullptr) {
+        m_info->m_connected = false;
+        emit statusUpdated(m_info);
+    }
+}
+
+void MTec::onRequestStatus()
+{
+    m_info = new MTecInfo;
+
+    if (m_firstTimeout.isValid() == false) {
+        /* No timestamp for timeout of first telegram defined,
+         * -> first request has not yet been sent
+         * -> do it now */
+
+        /* Note: m_firstRequest is set to false in
+         * the onReceivedHoldingRegister function */
+        if (m_firstRequest == true) {
+            m_firstTimeout = QDateTime::currentDateTime();
+            m_firstTimeout = m_firstTimeout.addSecs(MTec::FirstConnectionTimeout);
+
+            /* Save original modbus timeout, will be set again
+             * after first response is received */
+            m_modbusTimeout = m_modbusMaster->timeout();
+
+            /* The M-Tec heatpump requires a longer timeout to
+             * start-up the modbus communication */
+            m_modbusMaster->setTimeout(MTec::FirstConnectionTimeout);
+        } else {
+            /* Set back original modbus timeout */
+            m_modbusMaster->setTimeout(m_modbusTimeout);
+        }
+    } else {
+        QDateTime now = QDateTime::currentDateTime();
+
+        if (m_firstTimeout.msecsTo(now) < MTec::FirstConnectionTimeout) {
+            /* Timeout of first request not yet reached
+             * -> return without sending another request */
+            return;
+        }
+    }
+
+    QUuid reqId{};
```

Comment:
You don't use the `reqId`, either use it for matching the request with the response, or leave it away.

Reply:
Done, left out for now